File: 0022-generate-parentheses/0022-generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        # approach: backtracking with stack.pop()
        
        result = []
        stack = []
        
        def helper(opening, closing):
            if closing == n:
                result.append(''.join(stack))
                return
            
            if opening < n:
                stack.append('(')
                helper(opening + 1, closing)
                stack.pop()
            
            if closing < opening:
                stack.append(')')
                helper(opening, closing + 1)
                stack.pop()
        
        helper(0, 0)
        return result
        

File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    List<String> result = new ArrayList<>();
    Stack<String> stack = new Stack<>();
    
    public List<String> generateParenthesis(int n) {
        // approach: backtracking using stack.pop()
        
        stack.push("(");
        backtrack(n, 1, 0);
        
        return result;
    }
    
    private void backtrack(int n, int numOpen, int numClosed) {
        if (numOpen == n && numClosed == n) {
            StringBuilder stringBuilder = new StringBuilder();
            stack.stream().forEach(p -> stringBuilder.append(p));
            
            result.add(stringBuilder.toString());
        }
        
        if (numOpen < n) {
            stack.push("(");
            backtrack(n, numOpen + 1, numClosed);
            stack.pop();
        }
        
        if (numClosed < numOpen) {
            stack.push(")");
            backtrack(n, numOpen, numClosed + 1);
            stack.pop();
        }
    }
}